trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,181 @@
+                dataset = datasets[tag]
+                model.update(dataset)
+                model.optimize(dataset)
+
+    @classmethod
+    def from_record(
+        cls, search_space: SP, acquisition_rule: AcquisitionRule[S, SP], record: Record[S]
+    ) -> AskTellOptimizer[S, SP]:
+        """Creates new :class:`~AskTellOptimizer` instance from provided optimization state.
+
+        :param search_space: The space over which to search.
+        :param acquisition_rule: The acquisition rule, which defines how to search for a new point
+            on each optimization step.
+        :param record: Optimization state record.
+        """
+        # here we are recovering previously saved optimization state
+        # there the model was already trained
+        # thus there is no need to train it again
+        fit_model = False
+        return cls(
+            search_space,
+            record.datasets,
+            record.models,
+            fit_model,
+            acquisition_rule,
+            record.acquisition_state,
+        )
+
+    def ask(self) -> TensorType:
+        """Suggests a point (or points in batch mode) to observe by optimizing the acquisition function.
+        If the acquisition is stateful, its state is saved.
+
+        :return: A :class:`TensorType` instance representing suggested point(s).
+        """
+        points_or_stateful = self._acquisition_rule.acquire(
+            self._search_space, self._datasets, self._models
+        )
+
+        if callable(points_or_stateful):
+            self._acquisition_state, query_points = points_or_stateful(self._acquisition_state)
+        else:
+            query_points = points_or_stateful
+
+        return query_points
+
+    def tell(self, new_data: Mapping[str, Dataset] | Dataset) -> None:
+        """Updates optimizer state with new data.
+
+        :param new_data: New observed data.
+        :raise ValueError: If keys in ``new_data`` do not match those in already built dataset.
+        """
+        if isinstance(new_data, Dataset):
+            new_data = {OBJECTIVE: new_data}
+
+        if self._datasets.keys() != new_data.keys():
+            raise ValueError(
+                f"new_data keys {new_data.keys()} doesn't "
+                f"match dataset keys {self._datasets.keys()}"
+            )
+
+        self._datasets = {tag: self._datasets[tag] + new_data[tag] for tag in new_data}
+
+        for tag, model in self._models.items():
+            dataset = self._datasets[tag]
+            model.update(dataset)
+            model.optimize(dataset)
+
+    def get_state(self) -> Record[S]:
+        """Collects the current state of the optimization, which includes datasets,
+        models and acquisition state (if applicable).
+
+        :return: An optimization state record.
+        """
+        return Record(
+            datasets=self._datasets, models=self._models, acquisition_state=self._acquisition_state
```

Comment:
Do we need to deepcopy the models and acquisition_state to avoid the record being modified by future ask/tell commands? 

If you think it would be useful to expose _both_ a deecopied record that can be used safely and a current state that is more efficient then maybe call the first `to_record()` and make the second a `current_state` `@property`?

Reply:
That's a good question, haven't really thought about it. It may be also relevant in the context of retracing, so i won't change anything in that regard just yet.